osc/fetch.py

```diff
@@ -258,19 +268,26 @@ def run(self, buildinfo):
                         # mark it for downloading from the API
                         self.__add_cpio(i)
                     else:
-                        # if the checksum of the downloaded package doesn't match,
-                        # delete it and mark it for downloading from the API
+                        # HDRMD5 MISMATCH EXPLANATION:
+                        # OBS builds the noarch packages once per binary arch.
+                        # Such noarch packages are supposed to be nearly identical across all build arches,
+                        # any discrepancy in the payload and dependencies is considered a packaging bug.
+                        # But to guarantee that the local builds work identically to builds in OBS,
+                        # using the arch-specific copy of the noarch package is required.
+                        # Unfortunatelly only one of the noarch packages gets distributed
+                        # and can be downloaded from a local mirror.
+                        # All other noarch packages are available through the OBS API only.
+                        # Since there is currently no information about hdrmd5 checksums of published noarch packages,
+                        # we download them, verify hdrmd5 and re-download the package from OBS API on mismatch.
```

Comment:
noarch are the most prominent examples, but the same can also happen for architecture depend packages. eg. when you have multiple candidates in a project or when someone is messing around with the source history or the release number handling in a way that it is not increasing.

Reply:
Thanks for the details, I've used part of your text in the comment explaining the situation.